File: Wdtc/src/main/java/org/wdt/wdtc/ui/HomeWindow.java
package org.wdt.wdtc.ui;

import com.jfoenix.controls.JFXButton;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import org.apache.log4j.Logger;
import org.wdt.wdtc.auth.User;
import org.wdt.wdtc.auth.accounts.Accounts;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.launch.LaunchGame;
import org.wdt.wdtc.manger.FileManger;
import org.wdt.wdtc.manger.GameFolderManger;
import org.wdt.wdtc.manger.VMManger;
import org.wdt.wdtc.ui.users.NewUserWindows;
import org.wdt.wdtc.utils.PlatformUtils;
import org.wdt.wdtc.utils.ThreadUtils;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.IOException;
import java.util.Objects;

public class HomeWindow {
    private static final Logger logmaker = WdtcLogger.getLogger(HomeWindow.class);
    private Launcher launcher = Launcher.getPreferredLauncher();

    public HomeWindow(Launcher launcher) {
        this.launcher = launcher;
    }

    public HomeWindow() {
    }


    private static JFXButton getSettingButton(Stage MainStage) {
        JFXButton setting = new JFXButton("设置");
        setting.setPrefSize(128, 46);
        setting.setOnMousePressed(event -> {
            if (event.isControlDown()) {
                try {
                    Runtime.getRuntime().exec(new String[]{"cmd", "/c", "start", FileManger.getWdtcConfig().getCanonicalPath()});
                    logmaker.info("配置目录" + FileManger.getWdtcConfig() + "已打开");
                } catch (IOException e) {
                    ErrorWindow.setErrorWin(e);
                }
            } else {
                try {
                    SettingWindow.setSettingWin(MainStage);
                } catch (IOException e) {
                    ErrorWindow.setErrorWin(e);
                }
            }
        });
        return setting;
    }

    private static void NoUser(Stage MainStage) {
        NewUserWindows windows = new NewUserWindows(MainStage);
        windows.setTitle("您还没有账户呢!");
        windows.setType(Accounts.AccountsType.Offline);
        windows.show();
    }

    public void setHome(Stage MainStage) {
        AnchorPane pane = new AnchorPane();
        WindwosSize windwosSize = new WindwosSize(MainStage);
        MainStage.setTitle(Consoler.getWindowsTitle());
        VBox Menu = new VBox();
        Menu.setPrefSize(128, 450);
        JFXButton home = new JFXButton("首页");
        home.setPrefSize(128, 46);

        JFXButton UserWindow = new JFXButton("修改账户");
        UserWindow.setPrefSize(128, 46);
        UserWindow.setOnAction(event -> {
            NewUserWindows windows = new NewUserWindows(MainStage);
            windows.setTitle("注册账户");
            windows.show();
        });

        JFXButton downgame = new JFXButton("下载游戏");
        downgame.setDisable(!PlatformUtils.isOnline());
        downgame.setPrefSize(128, 46);
        downgame.setOnAction(event -> NewDownloadWindow.SetWin(MainStage));

        JFXButton startgame = new JFXButton("选择版本");
        startgame.setLayoutY(69);
        startgame.setPrefSize(128, 46);
        startgame.setOnAction(event -> {
            VersionChoose choose = new VersionChoose(Objects.requireNonNullElseGet(launcher, GameFolderManger::new));
            choose.setWindow(MainStage);
        });

        JFXButton VersionSetting = new JFXButton("版本设置");
        VersionSetting.setPrefSize(128, 46);
        if (launcher != null) {
            VersionSetting.setDisable(false);
            VersionSettingWindows windows = new VersionSettingWindows(launcher, MainStage);
            VersionSetting.setOnAction(event -> windows.setWindow());
        } else {
            VersionSetting.setDisable(true);
        }

        JFXButton setting = getSettingButton(MainStage);

        JFXButton github = new JFXButton("GitHub");
        github.setPrefSize(128, 46);
        github.setOnAction(event -> {
            try {
                Runtime.getRuntime().exec(new String[]{"cmd", "/C", "start", "https://github.com/Wd-t/Wdtc"});
            } catch (IOException e) {
                ErrorWindow.setErrorWin(e);
            }
        });

        Label name = new Label("Wdtc\n" + VMManger.getLauncherVersion());
        name.setLayoutX(17.0);
        name.setLayoutY(161.0);
        Label readme = new Label("一个简单到不能再简单的我的世界Java版启动器");
        readme.setLayoutX(180.0);
        readme.setLayoutY(180.0);
        readme.getStyleClass().add("readme");
        JFXButton LaunchGameButton = new JFXButton();
        if (launcher != null) {
            LaunchGameButton.setText("启动游戏\n" + launcher.getVersionNumber());
        } else {
            LaunchGameButton.setText("当前无游戏版本");
        }
        LaunchGameButton.setPrefSize(227, 89);
        LaunchGameButton.setLayoutX(335);
        LaunchGameButton.setLayoutY(316);
        LaunchGameButton.getStyleClass().add("BackGroundWriteButton");
        AnchorPane.setBottomAnchor(LaunchGameButton, 30.0);
        AnchorPane.setRightAnchor(LaunchGameButton, 30.0);
        LaunchGameButton.setOnAction(event -> {
            if (launcher != null) {
                if (User.isExistUserJsonFile()) {
                    ThreadUtils.StartThread(() -> {
                        try {
                            LaunchGame launch = new LaunchGame(launcher);
                            LauncherGameWindow launcherGameWindow = new LauncherGameWindow(launch.getProcess());
                            launcherGameWindow.StartGame();
                        } catch (IOException e) {
                            ErrorWindow.setErrorWin(e);
                        }
                    }).setName("Launch Game");
                } else {
                    NoUser(MainStage);
                }
            } else {
                NewDownloadWindow.SetWin(MainStage);
            }
        });
        Menu.getChildren().addAll(home, UserWindow, downgame, startgame, VersionSetting, setting, github);
        Menu.getStyleClass().add("BlackBorder");
        AnchorPane.setTopAnchor(Menu, 0.0);
        AnchorPane.setBottomAnchor(Menu, 0.0);
        AnchorPane.setLeftAnchor(Menu, 0.0);
        pane.getChildren().addAll(Menu, LaunchGameButton);
        windwosSize.ModifyWindwosSize(pane, readme);
        Consoler.setStylesheets(pane);
        pane.setBackground(Consoler.getBackground());
        Scene scene = new Scene(pane, 600, 450);
        MainStage.setScene(scene);
        if (!User.isExistUserJsonFile()) {
            NoUser(MainStage);
        }
    }
}

File: ui/src/main/kotlin/org/wdt/wdtc/ui/window/HomeWindow.kt
package org.wdt.wdtc.ui.window

import com.jfoenix.controls.JFXButton
import javafx.event.EventHandler
import javafx.scene.Scene
import javafx.scene.input.MouseEvent
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.VBox
import javafx.stage.Stage
import kotlinx.coroutines.*
import org.wdt.wdtc.core.impl.launch.LaunchGame
import org.wdt.wdtc.core.openapi.auth.Accounts.AccountsType
import org.wdt.wdtc.core.openapi.auth.isExistUserJsonFile
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.game.preferredVersion
import org.wdt.wdtc.core.openapi.manger.GameDirectoryManger
import org.wdt.wdtc.core.openapi.manger.wdtcCache
import org.wdt.wdtc.core.openapi.utils.isOnline
import org.wdt.wdtc.core.openapi.utils.noNull
import org.wdt.wdtc.core.openapi.utils.openSomething
import org.wdt.wdtc.core.openapi.utils.runOnIO
import org.wdt.wdtc.ui.window.user.NewUserWindows

class HomeWindow {
	private var version = preferredVersion
	
	constructor(version: Version?) {
		this.version = version
	}
	
	constructor()
	
	suspend fun setHome(mainStage: Stage) {
		runOnJavaFx {
			mainStage.run {
				title = windowsTitle
				getSizeManger()
			}
		}
		val home = JFXButton("首页").apply {
			setPrefSize(128.0, 46.0)
		}
		val userWindow = JFXButton("修改账户").apply {
			setPrefSize(128.0, 46.0)
			onAction = EventHandler {
				NewUserWindows(mainStage).apply {
					title = "注册账户"
				}.run { show() }
			}
		}
		val downgame = JFXButton("下载游戏").apply {
			isDisable = !isOnline
			setPrefSize(128.0, 46.0)
			onAction = EventHandler { GameVersionListWindow().setWindowScene(mainStage) }
		}
		val startgame = JFXButton("选择版本").apply {
			layoutY = 69.0
			setPrefSize(128.0, 46.0)
			onAction = EventHandler {
				VersionChooseWindow(version ?: GameDirectoryManger.DEFAULT_GAME_DIRECTORY).run {
					setWindow(mainStage)
				}
			}
		}
		val versionSetting = JFXButton("版本设置").apply {
			setPrefSize(128.0, 46.0)
			isDisable = version.let {
				if (it != null) {
					onAction = eventHandler {
						VersionSettingWindow(it, mainStage).setWindow()
					}
					false
				} else {
					true
				}
			}
		}
		val setting = getSettingButton(mainStage)
		val github = JFXButton("GitHub").apply {
			setPrefSize(128.0, 46.0)
			onAction = EventHandler { openSomething("https://github.com/wdtve/Wdtc") }
		}
		val launchGameButton = JFXButton().apply {
			text = version.let { if (it != null) "启动游戏${it.versionNumber}" else "当前无游戏版本" }
			setPrefSize(227.0, 89.0)
			layoutX = 335.0
			layoutY = 316.0
			styleClass.add("BackGroundWriteButton")
			onAction = eventHandler {
				withContext(Dispatchers.Default) {
					launchGameButtonTask(mainStage, version)
				}
			}
			setBottomAnchor(30.0)
			setRightAnchor(30.0)
		}
		val menu = VBox().apply {
			setPrefSize(128.0, 450.0)
			styleClass.add("BlackBorder")
			setTopAnchor(0.0)
			setBottomAnchor(0.0)
			setLeftAnchor(0.0)
			children.addAll(
				home, userWindow, downgame, startgame, versionSetting, setting.await(), github
			)
		}
		val pane = AnchorPane().apply {
			background = wdtcBackground
			children.addAll(menu, launchGameButton)
			setStylesheets()
		}
		runOnJavaFx {
			Scene(pane, 600.0, 450.0).let {
				mainStage.scene = it
			}
		}
		if (!isExistUserJsonFile) {
			runOnJavaFx {
				noUser(mainStage)
			}
		}
	}
	
	companion object {
		private suspend fun getSettingButton(mainStage: Stage): Deferred<JFXButton> = coroutineScope {
			async {
				JFXButton("设置").apply {
					setPrefSize(128.0, 46.0)
					onMousePressed = EventHandler {
						launchOnJavaFx {
							getSettingButtonAction(it, mainStage)
						}
					}
				}
			}
		}
		
		private suspend fun getSettingButtonAction(event: MouseEvent, mainStage: Stage) {
			if (event.isControlDown) {
				openSomething(wdtcCache)
			} else {
				tryCatching {
					SettingWindow.setSettingWin(mainStage)
				}
			}
		}
		
		private suspend fun launchGame(mainStage: Stage, version: Version) {
			val tasks = runOnIO {
				LaunchGame(version.noNull(), TaskWindow.taskPool).getTaskList { setWin(it, "启动失败") }
			}
			runOnJavaFx {
				TaskWindow(mainStage, tasks).run {
					showTaskStage()
				}
			}
			
		}
		
		
		suspend fun launchGameButtonTask(mainStage: Stage, version: Version?) {
			version.also {
				if (it != null) {
					if (isExistUserJsonFile) {
						launchGame(mainStage, it)
					} else {
						runOnJavaFx {
							noUser(mainStage)
						}
					}
				} else {
					GameVersionListWindow().setWindowScene(mainStage)
				}
			}
		}
		
		private fun noUser(mainStage: Stage) {
			NewUserWindows(mainStage).apply {
				title = "您还没有账户呢!"
				type = AccountsType.OFFLINE
			}.show()
		}
	}
}
